amalgomate/modules.go

```diff
@@ -245,14 +250,57 @@ func copyModuleRecursively(modulePath, srcDir, dstDir string) error {
 // local directory specified in a "replace" directive in the "go.mod" file of the module in which "dir" is located, the
 // vendor directory of the module in which "dir" is located, etc.
 func moduleInfoForPackage(pkgName, dir string) (*GoModInfo, error) {
-	outputDirPkg, err := packageForPatternInDirectory(pkgName, dir, packages.NeedName|packages.NeedFiles)
+	// get package information from package name, which should include the module information
+	outputDirPkg, err := packageForPatternInDirectory(pkgName, dir, packages.NeedName|packages.NeedFiles|packages.NeedModule)
 	if err != nil {
 		return nil, err
 	}
 	if len(outputDirPkg.GoFiles) == 0 {
 		return nil, errors.Errorf("no Go files in package %s resolved from directory %s", pkgName, dir)
 	}
-	return moduleInfoForDirectory(filepath.Dir(outputDirPkg.GoFiles[0]))
+
+	if outputDirPkg.Module == nil {
+		return nil, errors.Errorf("unable to determine module for package %s resolved from directory %s", pkgName, dir)
+	}
+
+	// determine the module for the specified package (may differ from the package because the main package for the
+	// module may not be in the root directory of the module)
+	modulePath := outputDirPkg.Module.Path
+	if modulePath == "" {
+		return nil, errors.Errorf("could not determine module for package %q in directory %q", pkgName, dir)
+	}
+
+	// if resolved module has directory field, use it
+	if outputDirPkg.Module.Dir != "" {
+		return &GoModInfo{
+			Path: modulePath,
+			Dir:  outputDirPkg.Module.Dir,
+		}, nil
+	}
+
+	// use "go list -e -json" for the module in the specified directory. Do this to determine the module directory
+	// (which may be a resolved directory like the vendor directory). So far, this seems to be the only way to reliably
+	// determine the on-disk path to the module for a package that may not have the same import path as its module.
+	// Run with the "-e" flag because this command returns an error if the specified module path does not contain any
+	// ".go" files (even if it is a valid module root). However, in this failure mode the "Path" field is still
+	// populated correctly, which is the only information that is needed.
+	goListCmd := exec.Command("go", "list", "-e", "-json", modulePath)
```

Comment:
I believe you can also pass through a format template to extract just the info you are looking for rather than have to render json and unmarshal
```sh
go list -e -f '{{.Dir}}' <modulePath>
```

Reply:
Ah great point, particularly true if we're just trying to read out one field. Thanks!